wit/couchbase/transactions.wit

```diff
@@ -0,0 +1,148 @@
+package wasmcloud:couchbase@0.1.0-draft;
+
+/// Perform transactions that contain multiple operations, comprising:
+/// - Key-Value operations
+/// - SQL++ queries
+///
+/// Reference: https://docs.couchbase.com/go-sdk/current/howtos/subdocument-operations.html#retrieving
+interface transactions {
+  use types.{
+    document-id, request-span, document, subdocument-path, document-error, retry-strategy, mutation-metadata,
+    query-scan-consistency, query-profile-mode, collection
+  };
+  use document.{document-get-result};
+  use sqlpp.{sqlpp-value, sqlpp-query-status, sqlpp-query-metrics, sqlpp-query-warning};
+
+  /// A token representing a transaction (pseudo resource)
+  ///
+  /// NOTE: In future versions with more widespred WIT Resource support,
+  /// this transaction will turn into a transaction resource.
+  type tx-token = string;
+
+  /// Errors that occur when trying to create a transaction
+  variant create-tx-error {
+    /// A completely unexpected error
+    unexpected(string),
+  }
+
+  /// Create a transaction
+  create-tx: func() -> result<tx-token, create-tx-error>;
+
+  /// Errors that occur while performing an operation inside a given transaction
+  variant tx-error {
+    /// Transaction has expired
+    expired,
+    /// Transaction has (already) failed
+    failed,
+    /// Transaction failed after committing
+    failed-post-commit,
+    /// Invalid/unknown transaction
+    invalid-tx,
+    /// A completely unexpected error
+    unexpected(string),
+  }
+
+  /// Retrieve a document, inside a transaction
+  tx-document-get: func(
+    tx: tx-token,
+    collection: collection,
+    id: document-id
+  ) -> result<document-get-result, tx-error>;
+
+  /// Insert a document, inside a transaction
+  tx-document-insert: func(
+    tx: tx-token,
+    id: document-id,
+    document: document,
+  ) -> result<document-get-result, tx-error>;
+
+  /// Replace a document, inside a transaction
+  tx-document-replace: func(
+    tx: tx-token,
+    old: document,
+    new: document,
```

Comment:
I'm curious about why we need to pass `old` and `new` ? AFAICT, Couchbase SDKs do not require the old version of the document.

Reply:
So this is actually based on https://pkg.go.dev/github.com/couchbase/gocb/v2#TransactionAttemptContext.Replace -- it *seems* to require the contents of the original object to do a replace inside a transaction, am I understanding it wrong somehow here?

Reply:
Actually, I noticed the types are wrong here : when you perform a `get` within a transaction, it returns an object with tx metadata in it, and also the document itself. When doing a `replace` or a `remove`, you need to pass that object as the first argument. In the case of the replace, the second argument is indeed the new value for the document itself.

Reply:
Yeah, it's a bit weird -- https://pkg.go.dev/github.com/couchbase/gocb/v2#TransactionGetResult doesn't have that tx metadata but it's what is used for replace... 

What do you think we can do about this? Is this a bug that should be filed on the upstream SDK? When we implement w/ the Golang SDK, that's the one signature we're doing to be using right -- the one with `TransactionGetResult`

Reply:
`TransactionGetResult` internals are not exposed, as mentioned in the documentation you linked.

```go
type TransactionGetResult struct {
	// contains filtered or unexported fields
}
```

> contains filtered or unexported fields

This means you MUST use the object returned by a `tx.get`. The second argument is simply the document itself.

Reply:
In WIT, we don't have the same data hiding ability -- what a `document-get-result` returns is quite clear. Hopefully, that structure should be enough to perform the update.

Reply:
I can tell you right away that it won't be enough to perform the update.
If you want the details of the structure, you can check it out in the Javascript SDK : https://github.com/couchbase/couchnode/blob/12deee1c383900f65bfd4d70b63b8cf87ad18d86/lib/transactions.ts#L213

Reply:
> This can be handled on the side of the provider (the tx-token can carry the relevant metadata), so I think we should be able to use the result.

The `tx-token` is for the whole transaction. The metadata required are specific to the document

Reply:
Oh sorry what information is missing from the result to perform the update? Are you referring to `links` and `metadata`?

Reply:
Just to note what I was implying more explicitly -- given that all the transactions are correlated with a single transaction ID, it is possible for the provider to hold extra information regarding a single transaction (including links/metadata on a performed get). 

Honestly, we could write this without the `old` parameter at all, because the provider could easily save the results of a get it performed previously on the Golang side.

Reply:
Ah just saw https://github.com/couchbaselabs/wasmcloud-provider-couchbase/pull/2#discussion_r1658460819 -- yes, but inside the whole transaction, there is a serial process of operations right -- when a Get is performed inside a transaction, it is possible for the provider to save the results of that get (and easily correlatable via the document ID), and save the required metadata related to it.

This way, we don't have to expose internal structures (that are likely to change/we don't want people to actually try to use) in the WIT contract.

Reply:
That's fair.
The document itself does not contain its own key though, so the first argument should be the doc key, not the doc itself.
If this is abstracted away by the provider, now the issue will be that the dev could think that he can do a `tx-replace` by passing the id and the doc, without doing a `get` first. But the provider will not have the metadata for it and therefore won't be able to perform the request. 
The provider could perform the `get` itself, but I don't know.. I'm not super comfortable with it. WDYT ?

Reply:
Ah that's true, this is certainly a problem, thanks for explaining!

Do we add a document + ID requirement here? if the user can at least supply that we might be able to get past this, *without* exposing the internal-ish metadata  stuff

Reply:
> Do we add a document + ID requirement here?
No we have a tx, document-specific, metadata requirement.

In the current SDKs, it's either hidden - when offered by the language - or treated as an internal _blob_.

Reply:
We *could* use a `resource` type here, but I don't think that's ideal as it would make the actual data unnecessarily hard to access (and possibly take more round-trips to the provider) -- we'll have to treat operation results in transactions with an internal use field.

With this, we can have this function take the get result exactly, and be able to implement the correlation in the provider.

Reply:
Hey @JesusTheHun just added that here:

https://github.com/couchbaselabs/wasmcloud-provider-couchbase/pull/2/commits/757e34642b6b462805bff8e2fd1283c675fcf9bd

What do you think?

Reply:
To me, shadowing a `get` operation is sketchy. It can be convenient in some cases, but it can also introduce complexity. What about the doc is `get` later, do you give the copy you've fetch or do you fetch it again ? 

I understand that you want to abstract some stuff away, but balance benefits / risks doesn't feel favorable to me, in this case.

Reply:
> To me, shadowing a get operation is sketchy. It can be convenient in some cases, but it can also introduce complexity. What about the doc is get later, do you give the copy you've fetch or do you fetch it again ?

So I think this case in general can be handled on the provider side -- if in a transaction you fetch a document *twice*, we know what the ID of the document is and can correlate it. We can smooth that over for the user, though I'd expect the user to use the most recent version. 

The provider is going to have to keep some sort of transaction metadata lookup available for the length of the transaction in either case I think (which reflects how the current SDK is written).

Also, isn't this the same case with the Go SDK? What if someone does a get twice, which get result do they pass back?

> I understand that you want to abstract some stuff away, but balance benefits / risks doesn't feel favorable to me, in this case.

It's less about abstracting things away in this case and preventing performance degradation for dealing with transactions. The way WIT currently supports implementation/data hiding is with `resource`s, and those generally will require round-trips (because the caller has no idea what to do with the resource, it's completely opaque).

If we turn the get-result into a resource (and turn the data-getting into methods) then we might incur a round trip every time, when what we really want to do is save a bit of metadata along with that get (which we can do on the provider side).

Reply:
That said, I think this is something we can iterate on *pre* 0.1.0 -- if we find that the implementation is just too hard, and we have to turn it into a `resource`, so be it -- but I think the functionality is now equal w/ the metadata field.

Actually an *even simpler* way we could do this is pass back the document ID in `tx-document-get-result` -- we don't even need an opaque metadata ID. I'm definitely happy to make that change as well if you prefer it.

Reply:
@vados-cosmonic if feel like you really want to give life to your idea so let's go for it ! As you said, it's not even 1.0 so we have time to rectify things if we need to.

Reply:
Yeah! Let's see how it goes -- though I hope the implementation won't unearth anything new, I think we'll get a great chance to discuss with more concrete code in front of us (and seeing what usage is like!)